msal/src/main/java/com/microsoft/identity/client/CurrentTaskBrowserTabActivity.java

```diff
@@ -0,0 +1,148 @@
+import static com.microsoft.identity.common.adal.internal.AuthenticationConstants.AuthorizationIntentAction.REDIRECT_RETURNED_ACTION;
+
+
+/**
+ * MSAL activity class (needs to be public in order to be discoverable by the os) to get the browser redirect with auth code from authorize
+ * endpoint. This activity has to be exposed by "android:exported=true", and intent filter has to be declared in the
+ * manifest for the activity.
+ * <p>
+ * When the AuthorizationAgent is launched, and we're redirected back with the redirect
+ * uri (the redirect must be unique across apps on a device), the os will fire an intent with the redirect,
+ * and the BrowserTabActivity will be launched.
+ * <pre>
+ * &lt;intent-filter&gt;
+ *     &lt;action android:name="android.intent.action.VIEW" /&gt;
+ *
+ *     To receive implicit intents, have to put the activity in the category of default.
+ *     &lt;category android:name="android.intent.category.DEFAULT" /&gt;
+ *
+ *     The target activity allows itself to be started by a web browser to display data.
+ *     &lt;category android:name="android.intent.category.BROWSABLE" /&gt;
+ *
+ *     BrowserTabActivity will be launched when matching the custom url scheme.
+ *     &lt;data android:scheme="msalclientid" android:host="auth" /&gt;
+ * &lt;/intent-filter&gt;
+ * </pre>
+ */
+public final class CurrentTaskBrowserTabActivity extends Activity {
+    private static final String TAG = CurrentTaskBrowserTabActivity.class.getSimpleName();
+    private static final int REDIRECT_RECEIVED_CODE = 2;
+    private BroadcastReceiver mCloseBroadcastReceiver;
+    private int mTaskIdResponseFor;
+
+
+    @Override
+    protected void onCreate(final Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        final String response = getIntent().getDataString();
+        final HashMap<String, String> urlParameters = StringUtil.isEmpty(response) ? null : StringExtensions.getUrlParameters(response);
+        if(urlParameters != null){
+            if(urlParameters.containsKey("state")) {
+                mTaskIdResponseFor = AndroidTaskStateGenerator.getTaskFromState(decodeState(urlParameters.get("state")));
+            }else{
+                IllegalStateException ex = new IllegalStateException("Did not receive a task id in the authorization response.");
+                com.microsoft.identity.common.logging.Logger.error(TAG, "Did not receive task id in the authorization response", ex);
+                throw ex;
+            }
+        }
+
+        if (savedInstanceState == null
+                && getIntent() != null
+                && !StringUtil.isEmpty(getIntent().getDataString())) {
+            //Leaving this for now to set static response URI value
+            final Intent responseIntent = CurrentTaskBrowserAuthorizationFragment.createCustomTabResponseIntent(this, response);
```

Comment:
I assume this has a dependency on another common PR which is still in draft?

Reply:
Yes.  This PR isn't into dev.